javaagent-tooling/src/main/java/net/bytebuddy/agent/builder/AgentBuilderUtil.java

```diff
@@ -0,0 +1,340 @@
+                    }
+                    Set<String> loadingSuperTypes = DefineClassHandler.getSuperTypes();
+                    // super types set should contain at least java.lang.Object if this set is
+                    // empty something unexpected has happened, run all transformations
+                    if (loadingSuperTypes.isEmpty()) {
+                      return transformations.iterator();
+                    }
+                    for (String className : loadingSuperTypes) {
+                      // we know that this class is going to be transformed
+                      if (superTypeNames.contains(className)) {
+                        return transformations.iterator();
+                      }
+                    }
+
+                    // apply only the transformations that we can't decompose
+                    return unoptimizedTransformations.iterator();
+                  }
+
+                  return method.invoke(transformations, args);
+                });
+
+    agentBuilderTransformationsField.set(agentBuilder, list);
+  }
+
+  @Nullable
+  private static Result inspect(AgentBuilder.RawMatcher matcher) throws Exception {
+    if (matcher instanceof AgentBuilder.RawMatcher.Conjunction) {
+      List<AgentBuilder.RawMatcher> matchers = getDelegateMatchers(matcher);
+      if (!matchers.isEmpty()) {
+        // with our current matchers we only need to inspect the first element of the conjunction
+        return inspect(matchers.get(0));
+      }
+    } else if (matcher instanceof AgentBuilder.RawMatcher.ForElementMatchers) {
+      ElementMatcher<?> elementMatcher =
+          getDelegateMatcher((AgentBuilder.RawMatcher.ForElementMatchers) matcher);
+      Result result = inspect(elementMatcher);
+      if (result == null && logger.isLoggable(FINE)) {
+        logger.log(Level.FINE, "Could not decompose matcher {0}", elementMatcher);
+      }
+      return result;
+    }
+
+    return null;
+  }
+
+  @Nullable
+  private static Result inspect(ElementMatcher<?> matcher) throws Exception {
+    if (matcher instanceof DelegatingMatcher) {
+      Result result = inspect(((DelegatingMatcher) matcher).getDelegate());
+      if (matcher instanceof DelegatingSuperTypeMatcher) {
+        return Result.subtype(result);
+      }
+      return result;
+    } else if (matcher instanceof HasSuperClassMatcher) {
+      return Result.subtype(inspect(getDelegateMatcher((HasSuperClassMatcher<?>) matcher)));
+    } else if (matcher instanceof ErasureMatcher) {
+      return inspect(getDelegateMatcher((ErasureMatcher<?>) matcher));
+    } else if (matcher instanceof NameMatcher) {
+      return inspectNameMatcher((NameMatcher<?>) matcher);
+    } else if (matcher instanceof ElementMatcher.Junction.Conjunction) {
+      List<ElementMatcher<?>> matchers =
+          getDelegateMatchers((ElementMatcher.Junction.Conjunction<?>) matcher);
+      for (ElementMatcher<?> elementMatcher : matchers) {
```

Comment:
How will this behave with matchers like `named("A").or(named("B"))`? While we don't have these in our codebase, it is possible that users may have made custom extensions that do that.

Reply:
`or` is handled by `Disjunction` which we don't handle here, so it will get classified as "unoptimizable". If needed we could attempt to handle it by inspecting each component of the disjunction, if all of them yield a name then the end result will be the same as with `StringSetMatcher`. I replaced all the `named("A").or(named("B"))` constructs with `namedOneOf` to avoid implementing it.

Reply:
or is handled by Disjunction which we don't handle here, so it will get classified as "unoptimizable". If needed we could attempt to handle it by inspecting each component of the disjunction, if all of them yield a name then the end result will be the same as with StringSetMatcher. I replaced all the named("A").or(named("B")) constructs with namedOneOf to avoid implementing it.